src/annotator/range-util.ts

```diff
@@ -58,26 +51,27 @@ export function forEachNodeInRange(range, callback) {
   }
 }
 
+function nodeIsText(node: Node): node is Text {
+  const whitespaceOnly = /^\s*$/;
+  return (
+    node.nodeType === Node.TEXT_NODE && !node.textContent!.match(whitespaceOnly)
+  );
+}
+
 /**
  * Returns the bounding rectangles of non-whitespace text nodes in `range`.
  *
- * @param {Range} range
- * @return {Array<DOMRect>} Array of bounding rects in viewport coordinates.
+ * @return Array of bounding rects in viewport coordinates.
  */
-export function getTextBoundingBoxes(range) {
-  const whitespaceOnly = /^\s*$/;
-  const textNodes = /** @type {Text[]} */ ([]);
+export function getTextBoundingBoxes(range: Range): DOMRect[] {
+  const textNodes: Text[] = [];
   forEachNodeInRange(range, node => {
-    if (
-      node.nodeType === Node.TEXT_NODE &&
-      !(/** @type {string} */ (node.textContent).match(whitespaceOnly))
-    ) {
-      textNodes.push(/** @type {Text} */ (node));
+    if (nodeIsText(node)) {
+      textNodes.push(node);
     }
   });
 
-  /** @type {DOMRect[]} */
-  let rects = [];
+  let rects: DOMRect[] = [];
```

Comment:
Actually, it can be nicely refactored from `textNodes.forEach()` into `return textNodes.flatMap()`, with little changes, and all tests keep passing.

I have even tried to add changes just to make sure the tests failed in that case, so it looks good.

```diff
- let rects: DOMRect[] = [];
- textNodes.forEach(node => {
+ return textNodes.flatMap(node => {
  const nodeRange = node.ownerDocument.createRange();
  nodeRange.selectNodeContents(node);
  if (node === range.startContainer) {
    nodeRange.setStart(node, range.startOffset);
  }
  if (node === range.endContainer) {
    nodeRange.setEnd(node, range.endOffset);
  }
  if (nodeRange.collapsed) {
    // If the range ends at the start of this text node or starts at the end
    // of this node then do not include it.
-    return;
+    return [];
  }

  // Measure the range and translate from viewport to document coordinates
  const viewportRects = Array.from(nodeRange.getClientRects());
  nodeRange.detach();
-  rects = rects.concat(viewportRects);
+  return viewportRects;
});
- return rects;
```